port 'build -o' exporters from buildkit

ansemjo opened the pull request:
This is a ~first -- very rough --~ patch to port the build output flag from buildkit. I saw @AkihiroSuda's #225 and thought this would be a killer feature, so I tried my hand at it. It's mostly copy-pasted snippets from @tonistiigi's patch and the buildkit client.

Current issues:
- ~This patch does not properly handle all output `type=`s. Is support for all possible types necessary or should it be limited to `tar` and `local`?~
- I removed the requirement for the `-t tag` flag when `-o` is used because it does not make sense for `tar` or `local` output and is given in the form of a `name` field for images
- ~I had to comment the `Building ...`, `Setting up ...` and `Successfully built ...` prints because it gets in the way of tar output to stdout.~ The thing is, we only properly parse the output spec after the session is created, so we can't easily use the `stdoutUsed` approach or I don't yet see how without splitting the parser further. *edit:* I simply let them print to `stderr` since the buildkit output lands there already anyway.
- How do I properly attribute the copied functions? Tõnis is mentioned in the acknowledgements already.

~It's not meant for merging just yet but~ with this patch I can build quasi-arbitrary artifacts with a multi-stage Dockerfile, where the last stage `FROM scratch` only copies files of interest:

    ./img build -o type=tar,dest=artifacts.tar .

ansemjo commented:
I force-pushed a few changes. I think I'm handling all known exporter types now.

Docs / README probably need an update, too?

ansemjo commented:
Added a [section](https://github.com/genuinetools/img/pull/256/commits/e4db6bdebce9e6c36d654d02528c3f4759134ddb) to the README and looking at tests now.

All tests are run with the `doRun()` function at their core it seems? This function only returns combined stderr and stdout, which is a problem if -- for example -- I'd like to test wether the stdout is a valid tar archive and contains an expected file. If I leave that particular test out I could add a few tests for different exporter types or missing fields, etc.

ansemjo commented:
I've added a few tests now and `TestBuildOutputTarStdout` is failing because there is Buildkit output on Stdout. It works fine when running the built binary in a shell manually with:

    cat Dockerfile | ./img build - -o - | tar tvf -

Is there a flag on the session I need to explicitly set to tell the progress output to not use stdout?

AkihiroSuda commented on build.go:
```diff
@@ -424,5 +474,9 @@ func showProgress(ch chan *controlapi.StatusResponse, noConsole bool) error {
-	return progressui.DisplaySolveStatus(context.TODO(), "", c, os.Stdout, displayCh)
+	out := os.Stdout
+	if stdoutUsed {
+		out = os.Stderr
```
can be stderr permanently https://github.com/moby/buildkit/blob/4b7741dc29351b25649fac94fb7920c4bd0e29c9/cmd/buildctl/build.go#L292

ansemjo replied:
Nice, changed in d0a4dd70.

AkihiroSuda commented on build.go:
```diff
@@ -189,8 +219,8 @@ func (cmd *buildCommand) Run(args []string) (err error) {
-	fmt.Printf("Building %s\n", initialTag)
-	fmt.Println("Setting up the rootfs... this may take a bit.")
+	fmt.Fprintf(os.Stderr, "Building %s\n", initialTag)
+	fmt.Fprintln(os.Stderr, "Setting up the rootfs... this may take a bit.")
```
this is fine, but probably it is also possible to use progresswriter for these messages
https://github.com/docker/buildx/blob/3f18b659a09804c738226dbf6bacbcae54afd7c6/driver/docker-container/driver.go#L31

ansemjo replied:
I'm not entirely sure how I would implement that. Can we use `showProgress` with a fabricated StatusResponse with a single Vertex?

ansemjo commented:
The latest push fails in Travis, because I've switched to using the upstream `ParseOutput` func from `moby/buildkit/cmd/buildctl/build`, which is not vendored. What's the preferred solution? Update vendored files or copy relevant source files? The upstream version has updated the parsing a little and e.g. removed the special case for `-o -`.

Just running `make vendor` includes a lot of new crypto too:

```
git status
[...]
        vendor/github.com/moby/buildkit/cmd/
        vendor/github.com/moby/buildkit/session/secrets/secretsprovider/
        vendor/github.com/moby/buildkit/session/sshforward/sshprovider/
        vendor/golang.org/x/crypto/curve25519/
        vendor/golang.org/x/crypto/ed25519/
        vendor/golang.org/x/crypto/internal/
        vendor/golang.org/x/crypto/poly1305/
        vendor/golang.org/x/crypto/ssh/agent/
        vendor/golang.org/x/crypto/ssh/buffer.go
        vendor/golang.org/x/crypto/ssh/certs.go
        vendor/golang.org/x/crypto/ssh/channel.go
        vendor/golang.org/x/crypto/ssh/cipher.go
        vendor/golang.org/x/crypto/ssh/client.go
        vendor/golang.org/x/crypto/ssh/client_auth.go
        vendor/golang.org/x/crypto/ssh/common.go
        vendor/golang.org/x/crypto/ssh/connection.go
        vendor/golang.org/x/crypto/ssh/doc.go
        vendor/golang.org/x/crypto/ssh/handshake.go
        vendor/golang.org/x/crypto/ssh/kex.go
        vendor/golang.org/x/crypto/ssh/keys.go
        vendor/golang.org/x/crypto/ssh/mac.go
        vendor/golang.org/x/crypto/ssh/messages.go
        vendor/golang.org/x/crypto/ssh/mux.go
        vendor/golang.org/x/crypto/ssh/server.go
        vendor/golang.org/x/crypto/ssh/session.go
        vendor/golang.org/x/crypto/ssh/streamlocal.go
        vendor/golang.org/x/crypto/ssh/tcpip.go
        vendor/golang.org/x/crypto/ssh/transport.go
```

ansemjo commented:
Yay! :tada:
